AsyncDisplayKit/Details/Transactions/_ASAsyncTransaction.mm

```diff
@@ -47,9 +50,220 @@ - (NSString *)description
 
 @end
 
+// Lightweight operation queue for _ASAsyncTransaction that limits number of spawned threads
+class ASAsyncTransactionQueue
+{
+public:
+  
+  // Similar to dispatch_group_t
+  class Group
+  {
+  public:
+    // call when group is no longer needed; after last scheduled operation the group will delete itself
+    virtual void release() = 0;
+    
+    // schedule block on given queue
+    virtual void schedule(dispatch_queue_t queue, dispatch_block_t block) = 0;
+    
+    // dispatch block on given queue when all previously scheduled blocks finished executing
+    virtual void notify(dispatch_queue_t queue, dispatch_block_t block) = 0;
+    
+    // used when manually executing blocks
+    virtual void enter() = 0;
+    virtual void leave() = 0;
+    
+    // wait until all scheduled blocks finished executing
+    virtual void wait() = 0;
+    
+  protected:
+    virtual ~Group() { }; // call release() instead
+  };
+  
+  // Create new group
+  Group *createGroup();
+  
+  static ASAsyncTransactionQueue &instance();
+  
+private:
+  
+  struct GroupNotify
+  {
+    dispatch_block_t _block;
+    dispatch_queue_t _queue;
+  };
+  
+  class GroupImpl : public Group
+  {
+  public:
+    GroupImpl(ASAsyncTransactionQueue &queue)
+      : _pendingOperations(0)
+      , _releaseCalled(false)
+      , _queue(queue)
+    {
+    }
+    
+    virtual void release();
+    virtual void schedule(dispatch_queue_t queue, dispatch_block_t block);
+    virtual void notify(dispatch_queue_t queue, dispatch_block_t block);
+    virtual void enter();
+    virtual void leave();
+    virtual void wait();
+    
+    int _pendingOperations;
+    std::list<GroupNotify> _notifyList;
+    ASDN::Condition _condition;
+    BOOL _releaseCalled;
+    ASAsyncTransactionQueue &_queue;
+  };
+  
+  struct Operation
+  {
+    dispatch_block_t _block;
+    GroupImpl *_group;
+  };
+  
+  struct DispatchEntry // entry for each
+  {
+    std::list<Operation> _operations;
+    int _threadCount;
+  };
+  
+  std::map<dispatch_queue_t, DispatchEntry> _entries;
+  ASDN::Mutex _mutex;
+};
+
+ASAsyncTransactionQueue::Group* ASAsyncTransactionQueue::createGroup()
+{
+  Group *res = new GroupImpl(*this);
+  return res;
+}
+
+void ASAsyncTransactionQueue::GroupImpl::release()
+{
+  ASDN::MutexLocker locker(_queue._mutex);
+  
+  if (_pendingOperations == 0)  {
+    delete this;
+  } else {
+    _releaseCalled = true;
+  }
+}
+
+void ASAsyncTransactionQueue::GroupImpl::schedule(dispatch_queue_t queue, dispatch_block_t block)
+{
+  ASAsyncTransactionQueue &q = _queue;
+  ASDN::MutexLocker locker(q._mutex);
+  
+  DispatchEntry &entry = q._entries[queue];
+  
+  Operation operation;
+  operation._block = block;
+  operation._group = this;
+  entry._operations.push_back(operation);
+  
+  ++_pendingOperations; // enter group
+  
+  NSUInteger maxThreads = [NSProcessInfo processInfo].activeProcessorCount;
```

Comment:
Is there a reason to change this from the previously determined concurrent count value?  I think this should be at least *2, as render operations may block on loading an image from disk cache or other situations.